### PR TITLE
Prevent duplicate announcement

### PR DESCRIPTION
**Pull request by ali-turan7**
### Related issue:
https://github.com/dotnet/source-build/issues/3450

### Description:
Duplication has been avoided by determining whether or not an announcement with the identical title already exists. 

### Test:
https://dev.azure.com/dnceng/internal/_build/results?buildId=2190080&view=logs&j=19992227-62fb-5b50-4e29-3b72bd33eea1&t=a83d7a9e-2e19-5733-3923-52a07f576962
In addition, the code has been checked in the test repository.

### CODE REVIEW

File: eng/templates/stages/release.yml [hunk 1]
```diff
@@ -304,10 +302,22 @@ stages:
             echo "Announcement title: $title"
             echo "Announcement body: $body"
           else
-            echo "Submitting announcement."
-            announcement_url=$( gh api graphql -F RepoId=$(RepoId) -F categoryId=$(DiscussionCategoryId) -F body="$body" -F title="$title" -f query="$query" --template '{{.data.createDiscussion.discussion.url}}' )
+            query='query { repository(name: "${{ variables.announcementRepo }}", owner: "${{ variables.announcementOrg }}") { discussions(last: 100, categoryId: "$(DiscussionCategoryId)", orderBy: {field: UPDATED_AT, direction: DESC}) { edges { node { title url } } } } }'
```
[the remark this answers is not in the document]
Comment (ali-turan7) [reply eng/templates/stages/release.yml:306]: Done, Thank you :)

File: eng/templates/stages/release.yml [hunk 2]
```diff
@@ -304,10 +302,22 @@ stages:
...
+            if [[ -z "$discussion" || "$discussion" == "null" ]]; then
+              query='mutation($RepoId: ID!, $categoryId: ID!, $body: String!, $title: String!) { createDiscussion(input: {repositoryId: $RepoId, categoryId: $categoryId, body: $body, title: $title}) { discussion { url } } }'
+
+              echo "Submitting announcement."
+              announcement_url=$( gh api graphql -F RepoId=$(RepoId) -F categoryId=$(DiscussionCategoryId) -F body="$body" -F title="$title" -f query="$query" --template '{{.data.createDiscussion.discussion.url}}' )
+              echo "Announcement URL: $announcement_url"
+            else
+              existing_announcement_url=$(echo "$discussion" | jq -r '.url')
+              echo "##vso[task.logissue type=warning]Announcement already exists. Skipping submission."
+              echo "##vso[task.logissue type=warning]Existing Announcement URL: $existing_announcement_url"
```
[the remark this answers is not in the document]
Comment (ali-turan7) [reply eng/templates/stages/release.yml]: Done, Thank you :)

File: eng/templates/stages/release.yml [hunk 3]
```diff
@@ -304,10 +302,22 @@ stages:
             echo "Announcement title: $title"
             echo "Announcement body: $body"
           else
-            echo "Submitting announcement."
-            announcement_url=$( gh api graphql -F RepoId=$(RepoId) -F categoryId=$(DiscussionCategoryId) -F body="$body" -F title="$title" -f query="$query" --template '{{.data.createDiscussion.discussion.url}}' )
+            # Checking for an already existing announcement
+            query='query { repository(name: "${{ variables.announcementRepo }}", owner: "${{ variables.announcementOrg }}") { discussions(last: 100, categoryId: "$(DiscussionCategoryId)", orderBy: {field: UPDATED_AT, direction: DESC}) { edges { node { title url } } } } }'
+
+            response=$(gh api graphql -f query="$query")
+            discussion=$(echo "$response" | jq -r '.data.repository.discussions.edges[] | select(.node.title == "'"$title"'") | .node')
```
Comment (mthalman) [inline eng/templates/stages/release.yml:309]: Consider more descriptive variable names:

* `query` => `recentAnnouncementsQuery`
* `response` => `recentAnnouncementsResponse`
* `discussion` => `discussionWithSameTitle`

Comment (premun) [reply eng/templates/stages/release.yml]: I will put both steps - getting discussion category and submitting the announcement in a script that's just easy to test etc. This shouldn't take long

**Comment by premun** [comment]
Sample release: https://dev.azure.com/dnceng/internal/_build/results?buildId=2191272&view=logs&j=19992227-62fb-5b50-4e29-3b72bd33eea1&t=13c737b6-da78-5a0a-c5a4-3c44e112dce6&l=16

**Review by mthalman — APPROVED** [review]

**Pull request merged**